components/brave_news/browser/brave_news_p3a.cc

```diff
@@ -67,110 +44,265 @@ void RecordGeneralUsage() {
+  }
+}
+
+void NewsMetrics::RecordFeatureEnabledChange() {
+  bool enabled = IsNewsEnabled();
+  bool was_ever_enabled = prefs_->GetBoolean(prefs::kBraveNewsWasEverEnabled);
   if (!enabled && !was_ever_enabled) {
     // If the user clicked "no thanks" on the NTP, then we don't want to record
     // this as an opt-out, since they were never opted in.
     return;
   }
-  prefs->SetBoolean(prefs::kBraveNewsWasEverEnabled, true);
+  prefs_->SetBoolean(prefs::kBraveNewsWasEverEnabled, true);
   UMA_HISTOGRAM_BOOLEAN(kIsEnabledHistogramName, enabled);
 }
 
-void RecordAtInit(PrefService* prefs) {
-  RecordLastUsageTime(prefs);
-  RecordNewUserReturning(prefs);
+void NewsMetrics::RecordAtInit() {
+  RecordLastUsageTime();
+  RecordNewUserReturning();
+
+  RecordDirectFeedsTotal();
+  RecordWeeklyAddedDirectFeedsCount(0);
+  RecordWeeklySessionCount(false);
+  RecordWeeklyDisplayAdsViewedCount(false);
+
+  for (const auto action : kAllActionTypes) {
+    RecordTotalActionCount(action, 0);
+  }
+  RecordVisitCardDepth({});
+
+  if (IsNewsEnabled()) {
+    prefs_->SetBoolean(prefs::kBraveNewsWasEverEnabled, true);
+  }
+
+  report_timer_.Start(
+      FROM_HERE, base::Time::Now() + kReportInterval,
+      base::BindOnce(&NewsMetrics::RecordAtInit, base::Unretained(this)));
+}
+
+bool NewsMetrics::IsNewsEnabled() {
+  return prefs_->GetBoolean(prefs::kBraveNewsOptedIn) &&
+         prefs_->GetBoolean(prefs::kNewTabPageShowToday);
+}
 
-  RecordDirectFeedsTotal(prefs);
-  RecordWeeklyAddedDirectFeedsCount(prefs, 0);
-  RecordWeeklySessionCount(prefs, false);
-  RecordWeeklyDisplayAdsViewedCount(prefs, false);
-  RecordTotalCardViews(prefs, 0);
+bool NewsMetrics::IsMonthlyActiveUser() {
+  base::Time last_usage = prefs_->GetTime(prefs::kBraveNewsLastSessionTime);
+  return base::Time::Now() - last_usage < kMonthlyUserTimeThreshold;
+}
 
-  if (IsNewsEnabled(prefs)) {
-    prefs->SetBoolean(prefs::kBraveNewsWasEverEnabled, true);
+uint64_t NewsMetrics::AddToWeeklyStorageAndGetSum(const char* pref_name,
+                                                  int change) {
+  WeeklyStorage storage(prefs_, pref_name);
+  if (change > 0) {
+    storage.AddDelta(1);
+  } else if (change < 0) {
+    storage.SubDelta(1);
   }
+  return storage.GetWeeklySum();
+}
+
+void NewsMetrics::RecordLastUsageTime() {
+  p3a_utils::RecordFeatureLastUsageTimeMetric(
+      prefs_, prefs::kBraveNewsLastSessionTime, kLastUsageTimeHistogramName);
+}
+
+void NewsMetrics::RecordNewUserReturning() {
+  p3a_utils::RecordFeatureNewUserReturning(
+      prefs_, prefs::kBraveNewsFirstSessionTime,
+      prefs::kBraveNewsLastSessionTime, prefs::kBraveNewsUsedSecondDay,
+      kNewUserReturningHistogramName);
+}
+
+void NewsMetrics::RecordWeeklySessionCount(bool is_add) {
+  // Track how many times in the past week
+  // user has scrolled to Brave News.
+  constexpr int buckets[] = {0, 1, 3, 7, 12, 18, 25, 1000};
+  uint64_t total_session_count =
```

Comment:
It feels slightly weird passing a boolean into a method expecting an int - I guess it's fine here because we only want to either
1. Add one session
2. Do nothing

Reply:
correct. implicit conversion is fine here